### ISSUE TITLE
CMesh object needs to be an instance of Triangulation

### ISSUE THREAD

**Issue by danshapero** [danshapero/geometry-is-hard#2]
In order to enable interactive plotting, I've written a Cython extension class call `CMesh`, which wraps an equivalent C struct and has properties that allow Python to query the triangulation object. However, matplotlib calls `isinstance` in various functions for making plots over triangular meshes to check that the argument is an instance of the class `Triangulation`, for example [here](https://github.com/matplotlib/matplotlib/blob/4314d447dfc7127daa80fa295c9bd56cf07faf01/lib/matplotlib/tri/triangulation.py#L134).

I need to override the `__instancecheck__` magic method so that a `CMesh` object can pass for a `Triangulation`, but I don't know how to do this correctly.

**Comment by cswiercz** [comment]
@danshapero Some resources on `__instancecheck__`:

* [Python Documentation](https://docs.python.org/2/reference/datamodel.html#customizing-instance-and-subclass-checks)
* [Stackoverflow 13135712](http://stackoverflow.com/questions/13135712/class-method-instancecheck-does-not-work)

**Comment by cswiercz** [comment]
According to [this thread](https://groups.google.com/forum/#!topic/cython-users/8ffwBO4Gqq0) Python metaclasses are supported but not Cython metaclasses. (Cython metaclasses sounds like dark voodoo magic.) I *suppose* this makes sense given that Cython extenion types are Python classes with a light wrapper around an underlying C struct.

**Comment by cswiercz** [comment]
Do you have some example (attempted) plotting code?

**Comment by danshapero** [comment]
At one point I tried something like the code you wrote above, but to no avail. I think there was trouble invoking `isinstance(other, Triangulation)`, relating to whether `other` is expected to be an object or a type. I'll try again and make a new branch for it today so you can see.

[This thread](http://stackoverflow.com/questions/23901730/how-does-one-work-with-cdefined-metaclasses-in-cython) and some [Sage code](https://github.com/sagemath/sagelib/blob/master/sage/misc/classcall_metaclass.pyx) are more up-to-date on metaclasses in Cython.

**Comment by cswiercz** [comment]
Yeah, I have the following right now and it's not even printing the debug statements:

```python 
# cmake.pyx
...
class TriangulationInstance(type):
    def	__instancecheck__(self, other):
        print '\nself  =', self
        print '\nother =', other
        if isinstance(other, Triangulation):
            return True
        return isinstance(self, other)
...
```
```python
# test.py
from cmesh import CMesh
from matplotlib.tri import Triangulation
M = CMesh('meshes/example.1')
print isinstance(M,Triangulation)
```
```
False
```

**Comment by cswiercz** [comment]
Oh!!! I've been thinking about it completely backwards. The example from Stackoverflow creates a class `EnumInt` where, no matter what object is provided to `isinstance()`, it's evaluated to `True`. That is

```python
isinstance(<anything you want>, EnumInt) == True
```

However, that's not what we want in this case. We don't have any control over `Triangulation.__instancecheck__()` so some other method for figuring out how to make `isinstance(<instance of CMesh>, Trangulation)` evaluate to `True` is needed.

**Comment by cswiercz** [comment]
@danshapero So here's a dumb question: *why not make `CMesh` inherit from `Triangulation`?* You've set it up so that the duck typing approach would work if matplotlib didn't type-check. Why not go the rest of the way?
